### DISCUSSION TITLE
On Normal Closure the app doesn't retry even when shouldRetry set to true

### DISCUSSION THREAD

**Discussion by OttlikG** [enisdenjo/graphql-ws#449] [Q&A] [answered]
**Screenshot**

![218757654-5f46e4c6-a05e-4a13-9079-552ca45502ea](https://user-images.githubusercontent.com/2817762/218785828-83e444e2-767f-419d-9e41-7a4f9f5a140d.png)

**Expected Behaviour**
I'm expecting the connection return to normal after the server drops cause of restart.

**Actual Behaviour**
I set the following options:
```
{
  shouldRetry: () => true,
  retryAttempts: 50,
}
```
to `createClient` but the retry mechanism doesn't kicks in.
As I see the process doesn't get to `shouldRetryConnectOrThrow` so the shouldRetry configuration doesn't get called.

**Debug Information**
I run Hasura graphql server which throws the following error on Kubernetes pod restart.
```
[
    {
        "extensions": {
            "path": "$",
            "code": "postgres-error"
        },
        "message": "connection error"
    }
]
```
The socket.onclose function gets called when server restarts without the retry functionality.
The event type is `Normal Closure`.

If any more information is needed, let me know.

**Comment by enisdenjo** [comment, marked as answer]
Hey there, close code `1000: Normal Closure` means we're good and done - it should not be used as a problematic close code. graphql-ws intentionally skips this code and will not retry it:

https://github.com/enisdenjo/graphql-ws/blob/eea18ce208abed12d3a9e61adc2c28d618cb3803/src/client.ts#L841-L844

Consider using `1001: Going Away`, or even `1012: Service Restart`, or `1013: Try Again Later`, as a better-fitting indicator that the server is being restarted.

**Reply by OttlikG** [reply]
Not sure if there is anything I can change. I think it's Hasura that close the connection with 1000 on server restart.
Will try to run by this issue at their Github and come back later.

**Comment by OttlikG** [comment]
As it turned out, preparing the backend to send the aforementioned codes as a response would be complicated. We're facing this issue when Kubernetes do a rolling deployment, and it's difficult to signal when it's not a Normal Closure. In the interim, we would like to patch this library in a way when we receive the `1000` code; it should be interpreted as a retry.
Do you have any suggestions on how to do that?

**Reply by enisdenjo** [reply]
You can look at this part: https://github.com/enisdenjo/graphql-ws/blob/eea18ce208abed12d3a9e61adc2c28d618cb3803/src/client.ts#L841-L844.

An alternative would be a custom WebSocket implementation that augments the close codes so that it converts 1000 to any other erroneous close code.

**Comment by KayakinCoder** [comment]
Sorry for asking @enisdenjo before trying this out myself, but we're using Hasura also and I was wondering if `EventClosedListener` or something else could be used to potentially detect this situation on the client? Thanks

**Reply by enisdenjo** [reply]
Not really, when the closed event is emitted - its already too late - the connection is closed. You can do what mentioned [here](https://github.com/enisdenjo/graphql-ws/discussions/449#discussioncomment-5568441), provide a custom WebSocket implementation that changes the 1000 close code to something different.
